Add Lyric support

robert-hamilton36 opened the pull request:
Changes
Adds route for lyrics. Will show current lyric highlighted if saved file allows it
Add item detail page for individual songs that show the lyrics


Previews

#/Lyrics
![image](https://github.com/jellyfin/jellyfin-web/assets/69886955/1f7ac7b0-7bb2-4f40-8941-ca7697f97aeb)


#/details?id={songsitemId}
![image](https://github.com/jellyfin/jellyfin-web/assets/69886955/5c5eb124-fe93-4c27-97a8-9b49fc011482)

ebkalderon commented:
Hey there! Just wondering what other work might need to be done before this PR can be taken out of draft and reviewed. The implementation itself looks great. Might this be blocked on something that needs additional help shepherding through?

robert-hamilton36 commented:
Oh no, Sorry. I didn't mean to leave it in draft. I think its good to go and be reviewed.

thornbill commented:
I believe this will require changes to be compatible with https://github.com/jellyfin/jellyfin/pull/9951

dmitrylyzo commented on src/controllers/itemDetails/index.js:
```diff
@@ -41,6 +41,7 @@ import 'elements/emby-scroller/emby-scroller';
 import 'elements/emby-select/emby-select';
 
 import 'styles/scrollstyles.scss';
+import '../../styles/lyrics.scss';
```
I believe we can now use the absolute import path.
```suggestion
import 'styles/lyrics.scss';
import 'styles/scrollstyles.scss';
```

robert-hamilton36 replied:
Thanks! Styles aren't currently being used in itemDetails so this import has been removed (I forgot to remove it when I changed it)

Fixed [17c61f4](https://github.com/jellyfin/jellyfin-web/pull/4733/commits/17c61f4b45928a188276c88a0cc05b9d3a94e23a)

dmitrylyzo commented on src/controllers/itemDetails/index.js:
```diff
@@ -1069,6 +1072,38 @@ function enableScrollX() {
+            const itemsContainer = lyricContainer.querySelector('.itemsContainer');
+            if (itemsContainer) {
+                const html = response.Lyrics.reduce((htmlAccumulator, lyric) => {
+                    htmlAccumulator += lyric.Text + '<br/>';
```
Escape HTML `lyric.Text`?
Unless the lyric has HTML styling, but that's dangerous.

robert-hamilton36 replied:
fixed on [a1768dd](https://github.com/jellyfin/jellyfin-web/pull/4733/commits/a1768dd4df99aaa428d8ef4ebc03c4d288e837fc)

dmitrylyzo commented on src/controllers/itemDetails/index.js:
```diff
@@ -1306,7 +1383,6 @@ function renderChildren(page, item) {
     } else if (item.Type == 'MusicArtist') {
         query.SortBy = 'PremiereDate,ProductionYear,SortName';
     }
-
```
I would leave this empty line for readability.

robert-hamilton36 replied:
fixed on [f84ca77](https://github.com/jellyfin/jellyfin-web/pull/4733/commits/f84ca7711f789cb4065adef576e90f7117436792)

dmitrylyzo commented on src/controllers/lyrics.js:
```diff
@@ -0,0 +1,236 @@
+let isDynamicLyric = false;
+
+function dynamicLyricHtmlReducer(htmlAccumulator, lyric, index) {
+    htmlAccumulator += `<div class="lyrics dynamicLyric" id="lyricPosition${index}" data-lyrictime="${lyric.Start}">${lyric.Text}</div>`;
```
Again, shouldn't we escape HTML `lyric.Text`? Here and below.

robert-hamilton36 replied:
fixed on [a1768dd](https://github.com/jellyfin/jellyfin-web/pull/4733/commits/a1768dd4df99aaa428d8ef4ebc03c4d288e837fc)

dmitrylyzo commented on src/controllers/playback/queue/index.html:
```diff
@@ -81,7 +81,11 @@ <h2 class="nowPlayingPageTitle"></h2>
+                            <span class="material-icons mic_external_on" aria-hidden="true"></span>
+                        </button>
+
+                        <button is="paper-icon-button-light" class="btnShuffleQueue autoSize" title="${Shuffle}"> 
```
```suggestion
                        <button is="paper-icon-button-light" class="btnShuffleQueue autoSize" title="${Shuffle}">
```

dmitrylyzo commented on src/controllers/playback/queue/index.html:
```diff
@@ -81,7 +81,11 @@ <h2 class="nowPlayingPageTitle"></h2>
                         </button>
 
-                        <button is="paper-icon-button-light" class="btnShuffleQueue autoSize" title="${Shuffle}">
+                        <button is="paper-icon-button-light" class="btnLyrics autoSize hide" title="${Lyrics}"> 
```
```suggestion
                        <button is="paper-icon-button-light" class="btnLyrics autoSize hide" title="${Lyrics}">
```

dmitrylyzo commented on src/scripts/deleteHelper.js:
```diff
@@ -1,46 +1,52 @@
 
 export default {
-    deleteItem: deleteItem
+    deleteItem: deleteItem,
+    deleteLyrics: deleteLyrics
```
```suggestion
    deleteItem,
    deleteLyrics
```

dmitrylyzo reviewed (commented):
I reviewed without testing.

robert-hamilton36 commented on src/components/itemContextMenu.js:
```diff
@@ -173,6 +173,14 @@ export function getCommands(options) {
             });
         }
+        if (item.Type === 'Audio' && item.HasLyrics && window.location.href.includes(item.Id)) {
+            console.log('affirmative');
```
Thanks, apologies

robert-hamilton36 commented on src/components/itemContextMenu.js:
```diff
@@ -616,10 +640,17 @@ function editItem(apiClient, item) {
+    });
+}
+
+function deleteLyrics(apiClient, item) {
+    return new Promise(function (resolve, reject) {
+        import('../scripts/deleteHelper').then((deleteHelper) => {
+            deleteHelper.deleteLyrics(item).then(function () {
                 resolve(true);
             }, reject);
         });
```
Yeah this works. Should I also update line 640  deleteItems (which this was based on) to the same format?

robert-hamilton36 commented on src/components/nowPlayingBar/nowPlayingBar.js:
```diff
@@ -81,6 +85,8 @@ function getNowPlayingBarHtml() {
 
     html += '<button is="paper-icon-button-light" class="btnAirPlay mediaButton"><span class="material-icons airplay" aria-hidden="true"></span></button>';
 
+    html += '<button is="paper-icon-button-light" class="openLyricsButton mediaButton"><span class="material-icons mic_external_on" aria-hidden="true"></span></button>';
```
Sweet. I will do the whole file as well on a different branch. I'll add the title for this button on this branch however

robert-hamilton36 replied:
single title for lyrics added here [9e51b5c](https://github.com/jellyfin/jellyfin-web/pull/4733/commits/9e51b5c0ddead78a2b1bc41133b5417b45b37b8b)

robert-hamilton36 commented on src/components/itemContextMenu.js:
```diff
@@ -616,10 +640,17 @@ function editItem(apiClient, item) {
         import('../scripts/deleteHelper').then((deleteHelper) => {
-            deleteHelper.deleteItem({
-                item: item,
-                navigate: false
```
In most cases I believe this caused two navigations.

Navigation was also being handled by the methods that called itemContextMenu.show().

Although I think this change might be better off in a different branch, I'm looking at. So I will revert this.

On that topic, scripts/deleteHelper.js is only used in itemContextMenu.js. Is this something we should consider moving entirely into itemContextMenu instead of having a separate script

robert-hamilton36 commented on src/controllers/itemDetails/index.html:
```diff
@@ -187,6 +187,18 @@ <h2 class="sectionTitle sectionTitle-cards padded-right"></h2>
 
+                <div id="lyricsSection" class="verticalSection-extrabottompadding detailVerticalSection lyricsContainer hide">
+                    <h2 class="sectionTitle sectionTitle-cards padded-right">${Lyrics}</h2>
+                    <div is="emby-itemscontainer" class="vertical-list focuscontainer-x itemsContainer"></div>
```
Ok right, I think focuscontainer-x must have been a remnant from which ever section I decided to copy for a skeleton. I never managed to work out what it did, So kept it in in case it was important. I will remove it.

robert-hamilton36 replied:
Fixed on [063f16a](https://github.com/jellyfin/jellyfin-web/pull/4733/commits/063f16aaae784a19bb4dd7606ae066da89818df3)

thornbill commented:
You also need to rebase or merge master back into this branch to fix the conflicts. We are still interested in getting this in for 10.9.0 if we can get it reviewed and any issues resolved this week.

dmitrylyzo commented on src/controllers/itemDetails/index.js:
```diff
@@ -1115,11 +1149,53 @@ function renderMoreFromSeason(view, item, apiClient) {
+
+        const query = {
+            ParentId: item.AlbumId,
+            Fields: 'ItemCounts,PrimaryImageAspectRatio,BasicSyncInfo,CanDelete,MediaSourceCount',
```
AFAIK, this field has been removed.
```suggestion
            Fields: 'ItemCounts,PrimaryImageAspectRatio,CanDelete,MediaSourceCount',
```

dmitrylyzo commented on src/controllers/lyrics.html:
```diff
@@ -0,0 +1,6 @@
+        
+    </div>
+</div>
+    
```
Remove trailing spaces.

dmitrylyzo reviewed (commented):
Another pass.

1. Shouldn't we close the lyrics page after music stops?

2. This is not good in TV display mode as there is no scrolling. In fact, we don't handle text blocks in TV display mode well.

3. In Mobile display mode, the `Lyrics` button is small. This may be improved in the future. Maybe after a redesign or so.
![lyrics-small-button](https://github.com/jellyfin/jellyfin-web/assets/56478732/840ac490-0514-4e65-bf3a-5b482cca6da4)

robert-hamilton36 commented:
> Another pass.
> 
>     1. Shouldn't we close the lyrics page after music stops?
> 
>     2. This is not good in TV display mode as there is no scrolling. In fact, we don't handle text blocks in TV display mode well.
> 
>     3. In Mobile display mode, the `Lyrics` button is small. This may be improved in the future. Maybe after a redesign or so.
>        ![lyrics-small-button](https://private-user-images.githubusercontent.com/56478732/316880558-840ac490-0514-4e65-bf3a-5b482cca6da4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0ODIyOTUsIm5iZiI6MTcxMTQ4MTk5NSwicGF0aCI6Ii81NjQ3ODczMi8zMTY4ODA1NTgtODQwYWM0OTAtMDUxNC00ZTY1LWJmM2EtNWI0ODJjY2E2ZGE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI2VDE5Mzk1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIwMTI2ZTYwM2YzNTI3N2Y2N2RmN2RiZjg1YWY0YzllYThjYzljYWFkZjU0NWNhZDA0YTNmOWE3MjA2OGJjN2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1bXojUGQ-M3MNf5AotbfuqQf4JkUBu0zsjLNZ1uyqpc)
> 
> 
> UPD: 4. There is a new section called `More From This Album`. It doesn't seem to be related to the lyrics feature. Maybe it should be moved to the new PR?

1. Yes that would be a good idea. I'm thinking i'll just have to add an event for 'playbackstop' that appRouter.back()
2. Right silly me I guess I must have always used my scroll wheel, not realizing tvs couldn't do that. Assuming the keyboards arrow keys act the same as the tvs remotes navigation, I'll look into add scrolling. I'm not familiar with how its been implemented already in menus
3. Hmm yes, secondary buttons are hidden in mobile mode. I didn't realize and left the lyric button visible. I'll hide it, the problem is getting back onto that lyric page. For now, what I can do is add that lyric button to the item details page like so


![Add Mic](https://github.com/jellyfin/jellyfin-web/assets/69886955/fefdea13-bcea-4701-9ae0-d04f8cefe37b)

Which would require accessing view lyrics from the context menu and than clicking the mic button

4. Yes its not directly related to the lyric feature. My thinking was, that the new section is only relevant because of the new route #/details?id={songsitemId} that is opened on the view lyrics context menu. But will move into a seperate PR

robert-hamilton36 commented:
> Another pass.
> 
>     1. Shouldn't we close the lyrics page after music stops?
> 
>     2. This is not good in TV display mode as there is no scrolling. In fact, we don't handle text blocks in TV display mode well.
> 
>     3. In Mobile display mode, the `Lyrics` button is small. This may be improved in the future. Maybe after a redesign or so.
>        ![lyrics-small-button](https://private-user-images.githubusercontent.com/56478732/316880558-840ac490-0514-4e65-bf3a-5b482cca6da4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE1MTcxNTksIm5iZiI6MTcxMTUxNjg1OSwicGF0aCI6Ii81NjQ3ODczMi8zMTY4ODA1NTgtODQwYWM0OTAtMDUxNC00ZTY1LWJmM2EtNWI0ODJjY2E2ZGE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI3VDA1MjA1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEyYzQwNTM1ZTVjNzlmOTY4OGM4ZDUxOWQwNzkzMDYxZTQ2NzNlYjJhZTYxMmI1MGJkNTY3MjNlZWQ3NjgyNjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mTo9kmkdQzTnUj9nyn5q1TgYbniqka92DiX0cXpT-YE)
> 
> 
> UPD: 4. There is a new section called `More From This Album`. It doesn't seem to be related to the lyrics feature. Maybe it should be moved to the new PR?

1. fixed with https://github.com/jellyfin/jellyfin-web/pull/4733/commits/73c01fb688481c429f6d42092ec3ce3f2b6ff7f3
2. possible fix suggested above 
3. fixed with https://github.com/jellyfin/jellyfin-web/pull/4733/commits/cd8d5c73de2d10c9b350dadcbc67f4cfe75983f7 and then https://github.com/jellyfin/jellyfin-web/pull/4733/commits/83bee29bd7c375a8804b71b34091180ff49453b6 
4. fixed with https://github.com/jellyfin/jellyfin-web/pull/4733/commits/286e0bf4f8e82897d7d87f3ad5b80304bd50b464

robert-hamilton36 commented:
> You also need to rebase or merge master back into this branch to fix the conflicts. We are still interested in getting this in for 10.9.0 if we can get it reviewed and any issues resolved this week.

I hope this was it https://github.com/jellyfin/jellyfin-web/pull/4733/commits/f1d955a941ce069b322cad69525a1b4f497d786f

Or else I might need a bit of guidance

dmitrylyzo commented on src/components/playback/playbackmanager.js:
```diff
@@ -1640,6 +1640,7 @@ class PlaybackManager {
             }
 
             changeStream(player, ticks);
+            Events.trigger(player, 'seek', [ticks]);
```
Ideally, we should receive this event after the seeking is complete.
I believe that `timeupdate` is the correct event.

robert-hamilton36 commented:
I have found an issue. `realeaseCurrentPlayer()` doesn't seem to be unbinding events. So even after you exit the lyrics page the events are still being run in the background

dmitrylyzo commented on src/controllers/lyrics.js:
```diff
@@ -0,0 +1,239 @@
+        }
+    }
+
+    function getLyrics (serverId, itemId) {
```
```suggestion
    function getLyrics(serverId, itemId) {
```

dmitrylyzo commented on src/controllers/lyrics.js:
```diff
@@ -0,0 +1,239 @@
+        });
+    }
+
+    function bindToPlayer (player) {
```
```suggestion
    function bindToPlayer(player) {
```

dmitrylyzo commented on src/controllers/lyrics.js:
```diff
@@ -0,0 +1,239 @@
+    function onTimeUpdate(_, ticks) {
+        ticks *= TICKS_PER_SECOND;
+        if (isDynamicLyric) {
+            const currentIndex = getLyricIndex(ticks, savedLyrics);
```
What's the problem with `getCurrentPlayTime()`?
I don't think we should modify the `timeupdate` event for no reason. _In fact, it has data that differs from player to player - not good, imo._

dmitrylyzo reviewed (commented):
The only thing that bothers me is this behavior:
1. Start music.
2. Click the now playing bar (open queue).
3. Open lyrics.
4. Stop music using the now playing bar.
5. The lyrics page executes `appRouter.back`.
6. The queue page appears to be broken.

robert-hamilton36 commented:
> The only thing that bothers me is this behavior:
> 
>     1. Start music.
> 
>     2. Click the now playing bar (open queue).
> 
>     3. Open lyrics.
> 
>     4. Stop music using the now playing bar.
> 
>     5. The lyrics page executes `appRouter.back`.
> 
>     6. The queue page appears to be broken.

Yes so I think appRouter.goHome() is the best option. Because if you stay on the lyric page, the arrow in the top left will still take you to the broken queue page

dmitrylyzo commented on src/controllers/lyrics.js:
```diff
@@ -0,0 +1,238 @@
+        }
+    }
+
+    function onPlaybackStop () {
+        appRouter.back();
+    }
```
The lyrics page will be kept until there is a next item and will be closed at the end of the playing queue.
```suggestion
    function onPlaybackStop (_, state) {
        if (!state.NextMediaType) {
            appRouter.back();
        }
    }
```

dmitrylyzo commented:
> Yes so I think appRouter.goHome() is the best option. Because if you stay on the lyric page, the arrow in the top left will still take you to the broken queue page

This workaround seems to work:
```diff
diff --git a/src/controllers/playback/queue/index.js b/src/controllers/playback/queue/index.js
index d95081b91..60c541bea 100644
--- a/src/controllers/playback/queue/index.js
+++ b/src/controllers/playback/queue/index.js
@@ -1,4 +1,5 @@
 import RemoteControl from '../../../components/remotecontrol/remotecontrol';
+import { appRouter, history } from 'components/router/appRouter';
 import { playbackManager } from '../../../components/playback/playbackmanager';
 import libraryMenu from '../../../scripts/libraryMenu';
 import '../../../elements/emby-button/emby-button';
@@ -35,8 +36,19 @@ export default function (view) {
     }
 
     view.addEventListener('viewshow', function () {
+        const player = playbackManager.getCurrentPlayer();
+
+        if (!player) {
+            if (history.length > 1) {
+                appRouter.back();
+            } else {
+                appRouter.goHome();
+            }
+            return;
+        }
+
         libraryMenu.setTransparentMenu(true);
-        bindToPlayer(playbackManager.getCurrentPlayer());
+        bindToPlayer(player);
         document.addEventListener('keydown', onKeyDown);
 
         if (remoteControl) {
```
I hope it doesn't break anything.

robert-hamilton36 commented:
> > Yes so I think appRouter.goHome() is the best option. Because if you stay on the lyric page, the arrow in the top left will still take you to the broken queue page
> 
> This workaround seems to work:
> 
> ```diff
> diff --git a/src/controllers/playback/queue/index.js b/src/controllers/playback/queue/index.js
> index d95081b91..60c541bea 100644
> --- a/src/controllers/playback/queue/index.js
> +++ b/src/controllers/playback/queue/index.js
> @@ -1,4 +1,5 @@
>  import RemoteControl from '../../../components/remotecontrol/remotecontrol';
> +import { appRouter, history } from 'components/router/appRouter';
>  import { playbackManager } from '../../../components/playback/playbackmanager';
>  import libraryMenu from '../../../scripts/libraryMenu';
>  import '../../../elements/emby-button/emby-button';
> @@ -35,8 +36,19 @@ export default function (view) {
>      }
>  
>      view.addEventListener('viewshow', function () {
> +        const player = playbackManager.getCurrentPlayer();
> +
> +        if (!player) {
> +            if (history.length > 1) {
> +                appRouter.back();
> +            } else {
> +                appRouter.goHome();
> +            }
> +            return;
> +        }
> +
>          libraryMenu.setTransparentMenu(true);
> -        bindToPlayer(playbackManager.getCurrentPlayer());
> +        bindToPlayer(player);
>          document.addEventListener('keydown', onKeyDown);
>  
>          if (remoteControl) {
> ```
> 
> I hope it doesn't break anything.


If you manually enter `/#/queue` into the browser.
The queue still has the playbar but with 3 added menus:
1. Navigation
2. Send message
3. Enter text

They don't seem to do anything, but I don't know if its still in use or not.
But this fix will stop you from getting to this menu.


![Queue](https://github.com/jellyfin/jellyfin-web/assets/69886955/a4f81ffe-672b-47a0-9816-a1640cdb9ed7)

dmitrylyzo commented:
> If you manually enter `/#/queue` into the browser.
> The queue still has the playbar but with 3 added menus:

When I tested, it redirected back if there was something open, or went to the home page when entered in a blank tab.

> But this fix will stop you from getting to this menu.

I haven't tested remote control. This page should be accessible from the `PlayOn` button -> Remote control.

robert-hamilton36 commented:
I've reverted `onPlaybackStop` Behaviour to `goHome()`, until a work around is found

dmitrylyzo commented on src/controllers/lyrics.js:
```diff
@@ -0,0 +1,247 @@
+
+    function onPlaybackStop () {
+        // TODO: switch to appRouter.back(), with fix to navigation to /#/queue. Which is broken when it has nothing playing
+        appRouter.goHome();
```
Have you tried playing a playlist? I believe when the song reaches the end, the player emits a stop event. This way, you will be taken to the home page each time.

You can prevent this by checking `state.NextMediaType` (https://github.com/jellyfin/jellyfin-web/pull/4733#discussion_r1542632714).

nielsvanvelzen commented on src/components/itemContextMenu.js:
```diff
@@ -510,6 +529,15 @@ function executeCommand(item, id, options) {
                 break;
+            case 'lyrics': {
+                if (options.isMobile) {
+                    appRouter.show('Lyrics');
```
```suggestion
                    appRouter.show('lyrics');
```

nielsvanvelzen commented on src/components/nowPlayingBar/nowPlayingBar.js:
```diff
@@ -212,6 +219,14 @@ function bindEvents(elem) {
+        if (lyricPageActive) {
+            appRouter.back();
+        } else {
+            appRouter.show('Lyrics');
```
```suggestion
            appRouter.show('lyrics');
```

nielsvanvelzen commented on src/components/nowPlayingBar/nowPlayingBar.js:
```diff
@@ -754,6 +790,7 @@ function refreshFromPlayer(player, type) {
 }
 
 function bindToPlayer(player) {
+    lyricPageActive = appRouter.currentRouteInfo.path === '/Lyrics';
```
Not the best way to check because the paths are case insensitive.

dmitrylyzo replied:
Quick fix. Lyrics toggling doesn't work.
```suggestion
    lyricPageActive = appRouter.currentRouteInfo.path.toLowerCase() === '/lyrics';
```

nielsvanvelzen commented on src/components/remotecontrol/remotecontrol.js:
```diff
@@ -769,6 +771,10 @@ export default function () {
             }
         });
+        context.querySelector('.btnLyrics').addEventListener('click', function () {
+            appRouter.show('Lyrics');
```
```suggestion
            appRouter.show('lyrics');
```

dmitrylyzo commented on src/components/nowPlayingBar/nowPlayingBar.js:
```diff
@@ -786,6 +823,8 @@ Events.on(playbackManager, 'playerchange', function () {
 bindToPlayer(playbackManager.getCurrentPlayer());
 
 document.addEventListener('viewbeforeshow', function (e) {
+    lyricPageActive = appRouter.currentRouteInfo.path === '/Lyrics';
```
Quick fix. Lyrics toggling doesn't work.
```suggestion
    lyricPageActive = appRouter.currentRouteInfo.path.toLowerCase() === '/lyrics';
```

dmitrylyzo commented on src/styles/lyrics.scss:
```diff
@@ -0,0 +1,31 @@
+.dynamicLyricsContainer {
+    display: flex;
+    flex-direction: column;
+    height: max-content;
```
This doesn't work on webOS 1.2 (old WebKit). You can reproduce the looking by removing `height: max-content;`.
The only way I found is to set `height` for the lyrics element.

dmitrylyzo reviewed (commented):
_Oops, I forgot to submit review._

dmitrylyzo commented on src/controllers/lyrics.html:
```diff
@@ -0,0 +1,4 @@
+<div id="lyricPage" data-role="page" class="page lyricPage" data-backbutton="true">
+    <div class="dynamicLyricsContainer padded-bottom-page">
+    </div>
```
For some reason `flex` inside another `flex` has the wrong height - the now playing bar overlaps the last line.
Placing it in another `div` fixes the problem.
```suggestion
    <div>
        <div class="dynamicLyricsContainer padded-bottom-page">
        </div>
    </div>
```

dmitrylyzo commented on src/controllers/lyrics.js:
```diff
@@ -0,0 +1,250 @@
+        }
+    }
+
+    function onPlaybackStop (_, state) {
```
```suggestion
    function onPlaybackStop(_, state) {
```

dmitrylyzo commented on src/controllers/lyrics.js:
```diff
@@ -0,0 +1,250 @@
+        }
+    }
+
+    view.addEventListener('viewshow', function() {
```
```suggestion
    view.addEventListener('viewshow', function () {
```

dmitrylyzo commented on src/controllers/lyrics.js:
```diff
@@ -0,0 +1,250 @@
+        }
+    });
+
+    view.addEventListener('viewbeforehide', function() {
```
```suggestion
    view.addEventListener('viewbeforehide', function () {
```

dmitrylyzo reviewed (commented):
I think these are the last ones. :crossed_fingers: 

This PR requires squashing before merging.

thornbill commented on src/components/nowPlayingBar/nowPlayingBar.js:
```diff
@@ -82,6 +86,8 @@ function getNowPlayingBarHtml() {
 
     html += `<button is="paper-icon-button-light" class="btnAirPlay mediaButton" title="${globalize.translate('AirPlay')}"><span class="material-icons airplay" aria-hidden="true"></span></button>`;
 
+    html += `<button is="paper-icon-button-light" class="openLyricsButton mediaButton" title="${globalize.translate('Lyrics')}"><span class="material-icons mic_external_on" aria-hidden="true"></span></button>`;
```
Why not use the lyrics icon?

dmitrylyzo replied:
`lyrics` icon conflicts with new `lyrics` class. But it looks better.

thornbill commented on src/controllers/playback/queue/index.html:
```diff
@@ -81,6 +81,10 @@ <h2 class="nowPlayingPageTitle"></h2>
                         </button>
 
+                        <button is="paper-icon-button-light" class="btnLyrics autoSize hide" title="${Lyrics}">
+                            <span class="material-icons mic_external_on" aria-hidden="true"></span>
```
Same here.

dmitrylyzo reviewed (approved):
LGTM.

Who will squash the commits?

robert-hamilton36 commented:
I can squash, assuming this is alright
`git reset --soft master`
`git push-f master LyricsSupport`

dmitrylyzo commented:
> I can squash, assuming this is alright
> `git reset --soft master`
> `git push-f master LyricsSupport`

I don't think this will collect all the changes because merge commits (instead of rebase) were used.

There may be an easier way, but I am used to do it this way (being on the PR branch):
1. Update the upstream.
`git fetch upstream`, where `upstream` is a remote pointing to the upstream repo of jellyfin-web.
2. Backup current branch. To compare with it later.
`git branch backup-rebase`
3. Run interactive rebase.
`git rebase -i upstream/master`.
In the rebase script, you must `pick` the first commit, others - `fixup`.
4. Fix conflicts if any. Then, `git rebase --continue`. Until rebase is complete.
You can abort rebase with `git rebase --abort`.
6. Compare the current branch with the backup to see if there are any unnecessary changes.
7. If it is fine, force push.
If you ruin your branch, you can re-checkout it from the backup.

robert-hamilton36 commented:
> > I can squash, assuming this is alright
> > `git reset --soft master`
> > `git push-f master LyricsSupport`
> 
> I don't think this will collect all the changes because merge commits (instead of rebase) were used.
> 
> There may be an easier way, but I am used to do it this way (being on the PR branch):
> 
>     1. Update the upstream.
>        `git fetch upstream`, where `upstream` is a remote pointing to the upstream repo of jellyfin-web.
> 
>     2. Backup current branch. To compare with it later.
>        `git branch backup-rebase`
> 
>     3. Run interactive rebase.
>        `git rebase -i upstream/master`.
>        In the rebase script, you must `pick` the first commit, others - `fixup`.
> 
>     4. Fix conflicts if any. Then, `git rebase --continue`. Until rebase is complete.
>        You can abort rebase with `git rebase --abort`.
> 
>     5. Compare the current branch with the backup to see if there are any unnecessary changes.
> 
>     6. If it is fine, force push.
>        If you ruin your branch, you can re-checkout it from the backup.

Awesome! Thanks so much for all your help

ebkalderon commented:
Very excited to see this feature finally land in the 10.9 release, and not a moment too soon! Thanks to all involved. :tada:
